write-fonts/src/tables/cmap.rs

```diff
@@ -238,6 +237,222 @@ impl Cmap {
+    fn cost(&self) -> usize {
+        // a segment always costs 4 u16s (end, start, delta_id, id_range_offset)
+        const BASE_COST: usize = 4 * u16::RAW_BYTE_LEN;
+
+        if self.id_delta.is_some() {
+            BASE_COST
+        } else {
+            // and if there is not a common id_delta, we also need to add an item
+            // to the glyph_id_array for each char in the segment
+            BASE_COST + self.len() * u16::RAW_BYTE_LEN
+        }
+    }
+
+    /// `true` if we can merge other into self (other must follow self)
+    fn can_merge(&self, next: &Self) -> bool {
+        assert!(self.start_ix < next.start_ix);
+        self.end_char as u32 + 1 == next.start_char as u32
+    }
+
+    /// Combine this segment with one that immediately follows it.
+    ///
+    /// The caller must ensure that the two segments are contiguous.
+    fn combine(&self, other: Format4Segment) -> Format4Segment {
+        assert_eq!(other.start_ix, self.end_ix + 1,);
+        Format4Segment {
+            start_ix: self.start_ix,
+            start_char: self.start_char,
+            end_char: other.end_char,
+            end_ix: other.end_ix,
+            id_delta: None,
+        }
+    }
+}
+
+impl<'a> Format4SegmentComputer<'a> {
+    fn new(mappings: &'a [(char, GlyphId)]) -> Self {
+        // ignore chars above BMP:
+        let mappings = mappings
+            .iter()
+            .position(|(c, _)| u16::try_from(*c as u32).is_err())
+            .map(|bad_idx| &mappings[..bad_idx])
+            .unwrap_or(mappings);
+        Self {
+            mappings,
+            seg_start: 0,
+            gids_in_order: false,
+        }
+    }
+
+    /// a convenience method called from our iter in the various cases where
+    /// we emit a segment.
+    ///
+    /// a 'seg_len' of 0 means start == end, e.g. a segment of one glyph.
+    fn make_segment(&mut self, seg_len: usize) -> Format4Segment {
+        // if start == end, we should always use a delta.
+        let use_delta = self.gids_in_order || seg_len == 0;
+        let start_ix = self.seg_start;
+        let end_ix = self.seg_start + seg_len;
+        let start_char = self.mappings[start_ix].0;
+        let end_char = self.mappings[end_ix].0;
+        let result = Format4Segment {
+            start_ix,
+            end_ix,
+            start_char,
+            end_char,
+            id_delta: self
+                .mappings
+                .get(self.seg_start)
+                .map(|(cp, gid)| gid.to_u32() as i32 - *cp as u32 as i32)
+                .filter(|_| use_delta),
+        };
+        self.seg_start += seg_len + 1;
+        self.gids_in_order = false;
+        result
+    }
+
+    /// Find the next possible segment.
+    ///
+    /// A segment _must_ be a contiguous range of chars, but we where such a range
+    /// contains subranges that are also contiguous ranges of glyph ids, we will
+    /// split those subranges into separate segments.
```

Comment:
This seems more complicated than just finding the next run more directly? - we've got various special cases that go away if we do that:

1. `if prev gids were ordered but this one isn't, end prev segment`
1. tracking `self.gids_in_order`
1. `if gids were not previously ordered but are now`
1. `done looping then create the last segment`

https://github.com/googlefonts/fontations/pull/1229/files avoids all this. Is there a reason this is better that I'm missing?